### ISSUE TITLE
Move database-specific tx validation rules to be part of the database abstraction

### ISSUE THREAD

**Issue by ttmc** [bigchaindb/bigchaindb#1870]
The following blog post gives an overview of all the things that get checked when checking to see if a transaction is valid:

https://blog.bigchaindb.com/what-is-a-valid-transaction-in-bigchaindb-9a1a075a9598

There are now some MongoDB-specific transaction validation rules (which will apply to both `asset.data` and `metadata` once the metadata full text search PR #1812 is merged). That is:

1, The names of keys must not begin with `$` and they must not contain `.` or the null character (Unicode code point U+0000).
2. If the JSON document is stored in a collection that is being indexed for full text search, and it contains a key named `"language"`, then the value must be one of the supported language codes. That’s because MongoDB uses the language code to help it index the text.

Since those validation rules are backend-database-specific, they should be added to the backend database abstraction layer. (RethinkDB doesn't have any such special validation rules right now, for example.) Those backend-datagase-specific transaction validity checks should be called as part of the "check transaction" logic.

Note: Currently, the MongoDB-specific transaction validation rules _are_ checked:  in the webserver code (i.e. Flask code), so they're only checked for transactions coming in via the HTTP API.

**Comment by ttmc** [comment]
I labelled this issue `tendermint` just now but it's a change that would also be useful for the 1.x series.

**Comment by sbellem** [comment]
I have very mixed feelings about this.

I totally agree about putting things that are database specific in the database specific component(s). So totally in favour of this issue from that point of view, and great call!

Where I am less sure, is whether we want to complicate the transaction model in such a way. Ideally, I would prefer to see a transaction model that is more generic, more independent of databases-specific constraints. In other words the transaction model could be kept as simple as possible such that it can be working on different databases regardless of the constraints of these DBs. This may be too idealistic but I think it may be worth to give some thoughts to this view as the benefits may be quite considerable in terms of portability, and possible future developments, in which different projects could easily re-use the transaction model against different databases without having to worry about database-specific validation particularities.

**Comment by ttmc** [comment]
I think it's okay to have things that every implementation must satisfy (i.e. as part of the transaction model spec), but to allow for implementation-specific deviations (such as those forced on us when we use MongoDB but not when we use RethinkDB).

**Comment by ttmc** [comment]
On the other hand...

Maybe we could create a transaction model that is "maximally constrained" so that it will work with all known implementations, as you suggest. That's definitely appealing!

My concern is, what if, for example, someone wants to use AwesomeNewDB in their new implementation of an IPDB-compliant server, and it won't allow emojis in Unicode strings? All the existing, stored transactions with emojis in the Unicode strings will prevent their new implementation from working.

Maybe there is some set of constraints that we can put on the transaction JSON (i.e. as part of the transaction spec) so that that 99% of all JSON-storing databases can store them without complaint?

**Comment by ttmc** [comment]
Related issue: #1766

**Comment by sbellem** [comment]
Well, here's a very rough sketch of one way we could think of it:

1. Let's start by identifying the limitations of crypto-conditions.
2. Could we find a way to have a transaction model that does not add constraints besides those of crypto-conditions?
3. If not, what are those new minimal constraints?
4. Could we then present a transaction model which has a very minimal set of constraints, such that it can easily used on different databases, etc.
5. Can we provide hooks, or methods, techniques, etc that would allow designers and implementers to "specialize" the transaction model to their needs?

In other words, the idea is to think of the transaction model in terms of 3 fundamental layers;

1. crypto-conditions
2. IPDB protocol transaction model (basic) layer
3. extensible transaction model, as wild as it can get, according to the needs of various communities

our job at the moment would stop at 2 + whatever is needed for the MongoDB/RethinkDB specifics as intially outlined in this issue

The "whatever is needed for the MongoDB/RethinkDB specifics" part could then be an exercise on how we can actually practically the above proposed phase 3.

These are just "brainstormed" ideas.

**Comment by ttmc** [comment]
As a concrete example of a constraint that is forced on us by our use of MongoDB:

In the value of the `metadata` field, if that value is a JSON object, then the key names (keys) cannot begin with the character `$`

As you see, this isn't related to crypto-conditions. It's just a JSON constraint that comes from MongoDB. Thanks MongoDB! (There are more.) Moreover, other JSON-storing databases might impose other constraints. I wouldn't be surprised if some restrict the set of allowed Unicode code points in string values. Others probably have constraints on JSON numbers, such as maximum integer size, or maximum floating-point precision (which is why I mentioned issue #1766 above).

I think what we want is a _maximal_ set of JSON constraints, such that any transaction satisfying those constraints will work with almost any JSON-storing database.

**Comment by ttmc** [comment]
I was curious so I went to the RethinkDB docs to see what kind of constraints/restrictions it imposes on the JSON it can store. There are many restrictions! I guess we never ran into them so didn't notice. They're listed at:

https://rethinkdb.com/limitations/

**Comment by ttmc** [comment]
I'm closing this issue because it was replaced by issue #1900
